dapr/clients/grpc/_request.py

```diff
@@ -42,72 +100,159 @@ def __init__(
-        else:
-            raise ValueError(f'invalid data type {type(data)}')
+        super(InvokeServiceRequest, self).__init__(())
+
+        self._content_type = content_type
+        self._http_verb = None
+        self._http_querystring: Dict[str, str] = {}
+
+        self.set_data(data)
+
+        # Set content_type to application/json type if content_type
+        # is not given and date is bytes or str type.
+        if not self.is_proto() and not content_type:
+            self.content_type = DEFAULT_JSON_CONTENT_TYPE
+
+    @property
+    def http_verb(self) -> Optional[str]:
+        """Gets HTTP method in Dapr invocation request."""
+        return self._http_verb
+
+    @http_verb.setter
+    def http_verb(self, val: Optional[str]) -> None:
+        """Sets HTTP method to Dapr invocation request."""
+        if val not in self.HTTP_METHODS:
+            raise ValueError(f'{val} is the invalid HTTP verb.')
+        self._http_verb = val
+
+    @property
+    def http_querystring(self) -> Dict[str, str]:
+        """Gets HTTP querystring as dict."""
+        return self._http_querystring
+
+    def is_http(self) -> bool:
+        """Return true if this request is http compatible."""
+        return hasattr(self, '_http_verb') and not (not self._http_verb)
 
     @property
-    def data(self) -> GrpcAny:
+    def proto(self) -> GrpcAny:
+        """Gets raw data as proto any type."""
         return self._data
 
+    def is_proto(self) -> bool:
+        """Returns true if data is protocol-buffer serialized."""
+        return hasattr(self, '_data') and self._data.type_url != ''
+
+    def pack(self, val: Union[GrpcAny, GrpcMessage]) -> None:
+        """Serializes protocol buffer message.
+
+        Args:
+            message (:class:`GrpcMessage`, :class:`GrpcAny`): the protocol buffer message object
+
+        Raises:
+            ValueError: message is neither GrpcAny nor GrpcMessage.
+        """
+        if isinstance(val, GrpcAny):
+            self._data = val
+        elif isinstance(val, GrpcMessage):
+            self._data = GrpcAny()
+            self._data.Pack(val)
+        else:
+            raise ValueError('invalid data type')
+
+    def unpack(self, message: GrpcMessage) -> None:
+        """Deserializes the serialized protocol buffer message.
+
+        Args:
+            message (:obj:`GrpcMessage`): the protocol buffer message object
+                to which the response data is deserialized.
+
+        Raises:
+            ValueError: message is not protocol buffer message object or message's type is not
+                matched with the response data type
+        """
+        unpack(self.proto, message)
+
+    @property
+    def data(self) -> bytes:
+        """Gets request data as bytes."""
+        if self.is_proto():
+            raise ValueError('data is protocol buffer message object.')
+        return self._data.value
+
+    @data.setter
+    def data(self, val: Union[str, bytes]) -> None:
```

Comment:
Why can't this have GrpcAny, GrpcMessage also?

Reply:
data getter returns bytes data so that accepting Grpc* object may not make sense.